File: DFS,BFS/programmers_ctest_DFSBFS_3.py
from typing import List

#도달 가능한 단어인지 확인해주는 함수
def ableToGo(s1:str,s2:str,length:int):
    diff=0
    for i in range(length):
        if s1[i]!=s2[i]:
            diff+=1
    
    return diff==1

def solution(begin, target, words):
    length=len(begin)
    
    words=[begin]+words   

    res=[]
    def dfs(node, step, visited:List):
        if node==target:
            res.append(step)
            return
        if node not in visited:
            visited.append(node)

        filtered=words.copy()
        filtered.remove(node)
        for item in filtered:
            if item not in visited and ableToGo(node,item,length):
                dfs(item,step+1,visited)
    
    dfs(begin,0, [])
    if not res:
        return 0
    return min(res)

File: DFS,BFS/programmers_ctest_DFSBFS_3.js
function canMove(s1,s2, length){
    let diff=0;
    for(let i=0; i<length; i++){
        if (s1[i]!==s2[i]){
            diff++;
        }
    }
    if (diff===1){
        return true
    }else{
        return false
    }
}

function solution(begin, target, words) {
    words=[begin].concat(words);
    const len=begin.length;
    const result=[];

    function dfs(node,step,visited){
        if(node===target){
            result.push(step);
            return;
        }
        if (!visited.includes(node))visited.push(node);

        const filtered=words.filter(x=>!visited.includes(x))
        
        for(let i=0;i<filtered.length;i++){
            if(canMove(node,filtered[i],len)){
                dfs(filtered[i],step+1,visited)
            }
        }
    }

    dfs(begin,0,[]);

    console.log(result)
    if (result.length===0){ return 0}
    return Math.min(...result)
}